drivers/volume/portworx/portworx.go

```diff
@@ -1778,6 +1781,287 @@ func (d *portworx) ValidatePureVolumesNoReplicaSets(volumeName string, params ma
+		}
+		currentDevices[n.MgmtIp] = localPathMap
+	}
+
+	return currentDevices, nil
+}
+
+// InitializePureLocalVolumePaths sets the baseline for how many Pure devices are attached to each node, so that we can compare against it later
+func (d *portworx) InitializePureLocalVolumePaths() error {
+	currentDevices, err := d.getCurrentPureLocalVolumePaths()
+	if err != nil {
+		return err
+	}
+	logrus.Infof("Pure device baseline initialized to: %+v", currentDevices)
+	d.pureDeviceBaseline = currentDevices
+
+	// TODO: we should only include non-FACD drives here. Eventually, we should handle FACD as its own kind of volume in ValidatePureLocalVolumePaths and validate it matches the pxctl cd l output
+
+	return nil
+}
+
+// ValidatePureLocalVolumePaths checks that the given volumes all have the proper local paths present, *and that no other unexpected ones are present*
+func (d *portworx) ValidatePureLocalVolumePaths() error {
+	t := func() error {
+		currentDevices, err := d.getCurrentPureLocalVolumePaths()
+		if err != nil {
+			return err
+		}
+		logrus.Infof("Current pure devices: %+v", currentDevices)
+
+		// TODO: handle FACD drive failovers properly (these will show up as new devices but they're actually fine)
+		// Remove all devices that are in the baseline (complex set math time!). Also warn if any baseline devices are now missing?
+		for node, baselineDevices := range d.pureDeviceBaseline {
+			currentNodeDevices := currentDevices[node]
+			for wwid := range baselineDevices {
+				if _, exists := currentNodeDevices[wwid]; exists {
+					delete(currentNodeDevices, wwid)
+				} else {
+					return fmt.Errorf("baseline FA device %s originally present on node %s is missing, this should not happen", wwid, node)
+				}
+			}
+		}
+
+		allVolNames, err := d.ListAllVolumes()
+		if err != nil {
+			return err
+		}
+		// Inspect all volumes provided to get the PX spec
+		fadaVolumes := []*api.Volume{}
+		for _, volName := range allVolNames {
+			v, err := d.InspectVolume(volName)
+			if err != nil {
+				return err
+			}
+			if !v.Spec.IsPureVolume() {
+				continue
+			}
+			if v.Spec.ProxySpec.PureBlockSpec == nil {
+				// TODO: handle FBDA volumes properly as well
+				continue
+			}
+			fadaVolumes = append(fadaVolumes, v)
+		}
+
+		// For each volume, check which nodes it should be on. Remove it from the list of devices on that node. Error if we can't find it.
+		for _, v := range fadaVolumes {
+			// Find the node this volume is on
+			var foundNode *node.Node
+			attachedOn := v.GetAttachedOn()
+			for _, n := range node.GetStorageDriverNodes() {
+				if n.MgmtIp == attachedOn { // TODO: RWX support
+					tempVar := n
+					foundNode = &tempVar
+					break
+				}
+			}
+			if foundNode == nil {
+				logrus.Infof("Volume %s is not attached to any node, skipping", v.Locator.Name)
+				continue
+			}
+
+			// Find the device for this volume
+			var device *pureLocalPathEntry
+			for _, deviceEntry := range currentDevices[foundNode.MgmtIp] {
```

Comment:
same here: for FADA volumes we should check if singlePaths exist.

Reply:
See my comment on the last one but yeah we could at least check that we have one single path, good point